File: leetcode/python/0070. Climbing Stairs.py
# Time complexity: O(N)
# Space complexity: O(N)


class Solution(object):
    def climbStairs(self, n):
        dp = [0 for _ in range(n+1)]
        return self.recur(dp, n)
    
    def recur(self, dp, n):
        if n <= 1:
            dp[n] = 1

        if dp[n] != 0:
            return dp[n]

        dp[n] = self.recur(dp,n-1) + self.recur(dp, n-2)
        
        return dp[n]

File: leetcode/java/0070. Climbing Stairs.java
package leetcode.java;

// Time complexity: O(N)
// Space complexity: O(N)

class Solution {
    public int climbStairs(int n) {
        int[] dp = new int[46];
        return this.helper(n, dp);
    }

    public int helper(int n, int[] dp) {
        if (n <= 1) {
            return 1;
        }

        if (dp[n] != 0) {
            return dp[n];
        }

        dp[n] = this.helper(n-1, dp) + this.helper(n-2, dp);

        return dp[n];
    }
}